### ISSUE TITLE
Code Refactoring - Part 3 - Integrate the Mars Man Models Library

### ISSUE THREAD

**Issue by saidghamra** [AgoraCloud/ui#66]
This issue is part 3 of the multi part process of refactoring the code base.

In this issue, @MarcDAFrame and @waleedza99 will go through the code base and update all the existing models to take advantage of the [@mars-man/models package](https://www.npmjs.com/package/@mars-man/models). This will be accomplished in tandem with #64.

**Comment by saidghamra** [comment]
This is Marc's priority list for integrating the package into the UI:

```
1. workspace home page
  - show deployments
  - sidebar
2. deployment create / edit form
3. create workplace form
4. first workplace form
5. wiki
6. workplace metrics
  - create new d3.js
7. workspace users
8. admin users
```

**Comment by MarcDAFrame** [comment]
# New Structure outline

```python
src/
    app/
        workspace/
            # everything pertaining to workspace
            forms/
                # all the forms within a workspace
                # - creating deployments
                # - creating / updating a workspace
            models/
                # workspace models
                # - WorkspacesModel, WorkspaceModel
                # - DeploymentImagesModel
                # - WorkspaceMetricsModel
            views/
                select.tsx # WorksplaceSelect component
                index.tsx # this will include the workspace pages
            children/
                # all things contained within a workspace
                Admin/

                Deployments/
                    ...
                Projects/
                    ...
                Shortcuts/
                    ...
                Wiki/
                    ...

```


---
---
---


# List of files and revisions 
## Format 
- N/A means don't worry about this file 
- new - `./path/to/move/file/to`

---
---
---

# Admin
All files here will be moved to `./src/app/Admin/`
- `./src/app/containers/admin/index.tsx`
    - new - `./src/app/Admin/views/index.tsx`
- `./src/app/containers/admin/routes.tsx`
    - the routes will be held in `./src/app/index.tsx`
        - there is no reason to seperate them
        - unless we seperate all the routes into seperate sections?
        - but I think that this will reduce confusion
    
- `./src/app/containers/admin/users.tsx`
    - new - `./src/app/Admin/views/users.tsx`
- `./src/app/containers/admin/home.tsx`
    - new - `./src/app/Admin/views/index.tsx`
- `./src/app/workspace/user/forms/index.ts`
    - admin user: delete / update / unverify user
    - new - `./src/app/Admin/forms/index.ts`

# Workspace
- `./src/app/containers/workspace/index.tsx`
    - new - `./src/app/Workspace/views/index.tsx`
    - includes the WorkspaceHome
        - showing deployments, side bar...
- `./src/app/containers/workspace/update-workspace.tsx`
    - new - `./src/app/Workspace/views/forms.tsx`
- `./src/app/containers/workspace/workspace-metrics.tsx`
    - new - `./src/app/Workspace/views/metrics.tsx`
- `./src/app/containers/workspace/first-workspace.tsx`
    - new - `./src/app/Workspace/views/first-workspace.tsx`
- `./src/app/workspace/forms/index.ts`
    - new - `./src/app/Workspace/forms/index.ts`
- `./src/app/workspace/forms/update-workspace.ts`
    - new - `./src/app/Workspace/forms/index.ts`
- `./src/app/workspace/forms/create-workspace.ts`
    - new - `./src/app/Workspace/forms/index.ts`
- `./src/app/components/forms/workspace/index.tsx`
    - has CreateWorkspace form view
    - new - `./src/app/Workspace/views/forms.tsx`
- `./src/app/workspace/index.ts`
    - N/A - export file
- `./src/app/workspace/images.ts`
    - new - `./src/app/workspace/model/DeploymentImages.ts`
- `./src/app/workspace/model.ts`
    - new - `./src/app/workspace/model/index.ts`
- `./src/app/workspace/metric.ts`
    - new - `./src/app/workspace/model/Metrics.ts`


## Workspace Admin
- `./src/app/containers/workspace/admin/index.tsx`
    - new - `./src/app/workspace/children/Admin/views/index.tsx`
- `./src/app/containers/workspace/admin/routes.tsx`
    - the routes will be held in `./src/app/index.tsx`
- `./src/app/containers/workspace/admin/users.tsx`
    - new - `./src/app/workspace/children/Admin/views/users.tsx`
    - AdminUsersPage will go in index.tsx
    - AdminUsersTable will go in users.tsx
- `./src/app/containers/workspace/admin/home.tsx`
    - new - `./src/app/workspace/children/Admin/views/index.tsx`
    - currently doesn't include anything useful
### Workspace Admin Users Models / Forms
- `./src/app/workspace/user/forms/form.ts`
- `./src/app/workspace/user/forms/permissions.tsx`
- `./src/app/workspace/user/forms/invite-user.ts`
- `./src/app/workspace/user/models/index.ts`
    - N/A - export file
- `./src/app/workspace/user/models/model.ts`
    - includes both WorkspaceUsers and AdminUsers model
    - new - `./src/app/workspace/children/Admin/models/users.ts`
    - new - `./src/app/Admin/models/users.ts`
- `./src/app/workspace/user/models/dialog.ts`
    - includes model for the dialog for permissions and editing user
- `./src/app/workspace/user/models/permission.ts`
    - includes both WorkspaceUsersPermission and AdminUsersPermission model
    - new - `./src/app/workspace/children/Admin/models/permissions.ts`
    - new - `./src/app/Admin/models/permissions.ts`



# Deployments
- `./src/app/containers/workspace/deployments/index.tsx`
- `./src/app/containers/workspace/deployments/proxy/index.tsx`
- `./src/app/containers/workspace/deployments/info/index.tsx`
- `./src/app/containers/workspace/deployments/metrics.tsx`
- `./src/app/containers/workspace/deployments/logs/index.tsx`
- `./src/app/containers/workspace/deployments/create-deployment.tsx`
- `./src/app/components/deployment-card.tsx`
- `./src/app/workspace/deployment/models/index.ts`
- `./src/app/workspace/deployment/models/metrics.ts`
- `./src/app/workspace/deployment/models/logs.ts`
- `./src/app/workspace/deployment/index.ts`
- `./src/app/workspace/deployment/create-deployment-form.ts`

# Projects
- `./src/app/containers/workspace/projects/index.tsx`
- `./src/app/containers/workspace/projects/create-project.tsx`
- `./src/app/containers/workspace/projects/lanes.tsx`
- `./src/app/workspace/project/index.ts`
- `./src/app/workspace/project/model.ts`
- `./src/app/workspace/project/create-project-form.ts`
- `./src/app/workspace/project/lane/index.ts`
- `./src/app/workspace/project/lane/create-lane-form.ts`
- `./src/app/workspace/project/lane/model.ts`
- `./src/app/workspace/project/lane/task/index.ts`
- `./src/app/workspace/project/lane/task/model.ts`
- `./src/app/workspace/project/lane/task/create-task-form.ts`
- `./src/app/components/kanban/board.tsx`
- `./src/app/components/kanban/board-column.tsx`
- `./src/app/components/kanban/board-item.tsx`
- `./src/app/components/kanban/card.tsx`



# Wiki
- `./src/app/workspace/wiki.ts`
- `./src/app/containers/workspace/wiki/list.tsx`
- `./src/app/containers/workspace/wiki/index.tsx`
- `./src/app/containers/auth.tsx`
- `./src/app/containers/user-profile.tsx`



# Components
- `./src/app/components/index.ts`
    - N/A - export file
- `./src/app/components/gauge-chart.tsx`
- `./src/app/components/paper.tsx`
- `./src/app/index.tsx`
- `./src/app/styles/theme.tsx`



# Auth
## [Refactor Authentication Pages #71](https://github.com/AgoraCloud/ui/issues/71)
- `./src/app/workspace/user/forms/auth/signup.ts`
- `./src/app/workspace/user/forms/auth/index.ts`
- `./src/app/workspace/user/forms/auth/change-password.ts`
- `./src/app/workspace/user/forms/auth/verify.ts`
- `./src/app/workspace/user/forms/auth/login.ts`
- `./src/app/workspace/user/forms/auth/forgot-password.ts`

# Revise
## Input
most inputs will have to be revised a bit, will probably mostly stay the same however

- `./src/app/components/inputs/index.tsx`
- `./src/app/components/inputs/text-field.tsx`
- `./src/app/components/inputs/select.tsx`
    - WorkspaceSelect will be moved to
        - new - `./src/app/workspace/views/components/select.tsx`
- `./src/app/components/inputs/context-menu.tsx`
- `./src/app/components/inputs/modal.tsx`
- `./src/app/components/inputs/buttons.tsx`

## Route Guards
- `./src/app/components/forms/route-guards/workspaces/index.tsx`
- `./src/app/components/forms/route-guards/workspaces/redirects.tsx`
- `./src/app/components/forms/route-guards/index.tsx`
- `./src/app/components/forms/route-guards/auth.tsx`
- `./src/app/components/forms/route-guards/user.tsx`

## Dialogs
- `./src/app/components/dialogs/index.tsx`
- `./src/app/components/dialogs/confirm-delete.tsx`
- `./src/app/components/dialogs/users.tsx`
- `./src/app/components/dialogs/base.tsx`
- `./src/app/components/dialogs/permissions.tsx`
- `./src/app/components/dialogs/invite-user.tsx`
- `./src/app/workspace/user/forms/confirm-delete.ts`

## Stores
- `./src/app/stores/index.ts`
- `./src/app/stores/ui-store.ts`
- `./src/app/stores/router-store.ts`
- `./src/app/stores/user-store.ts`
- `./src/app/stores/snackbar-store.ts`
- `./src/app/stores/root-store.ts`
- `./src/app/stores/admin-store.ts`
- `./src/app/stores/auth-store.ts`

### Workspace Store
the workspace store will be revised as to move much of the logic into the `WorkspacesModel`
- `./src/app/stores/workspaces-store.ts`


# Delete
- `./src/app/forms/base.ts`
- `./src/app/forms/index.ts`
- `./src/app/forms/validators.ts`
- `./src/app/base-model.ts`




# TBD
- `./src/app/components/menu/index.ts`
- `./src/app/components/menu/app-menu.tsx`
- `./src/app/components/menu/top-bar.tsx`
- `./src/app/components/menu/side-bar.tsx`
- `./src/app/components/table.tsx`
- `./src/app/components/snackbar-manager.tsx`
- `./src/app/components/forms/permissions.tsx`
- `./src/app/components/forms/workspace/projects.tsx`
- `./src/app/components/forms/workspace/deployments.tsx`
- `./src/app/components/forms/workspace/user.tsx`


# Probably OK
- `./src/app/constants/index.ts`
- `./src/app/constants/stores.ts`
- `./src/app/constants/permissions.ts`
- `./src/app/constants/events.ts`
- `./src/app/config.tsx`

**Comment by saidghamra** [comment]
October 4 Update:
- @waleedza99  is no longer blocked, @MarcDAFrame has committed the workspaces and deployments model and forms integrated with the mars-man/models package
- @waleedza99 will start working on integrating the mars-man/models package into projects, project boards and project tasks

**Comment by MarcDAFrame** [comment]
I've pushed a version that has workspaces and deployment models mostly done. So far I've redone the CreateDeploymentForm + page, Workspace Home.

I'm yet to
1. implement the  CreateDeploymentRepo (but this should be fairly straight forward, just have to think about the smartest way to do it)
2. Deployments List - this will have to come after the CreateDeployment as I need deployments to test this

the next steps are 
1. finialize the models
2. update the views (forms, pages...) to work with the new models 
./src/app/index.tsx has all the routes and will act as a checklist for all the pages that need to be revisited.

so far only two pages have been fixed, but I figure that the next ones will be easier seeing as I've addressed most of the minutiae already

![image](https://user-images.githubusercontent.com/12240832/135889409-3c76b63e-279b-4b20-8f2d-15f82d3f670a.png)

**Comment by MarcDAFrame** [comment]
![image](https://user-images.githubusercontent.com/12240832/135889549-5c06b673-4b57-4ec2-9b85-d7772393a3c8.png)

**Comment by MarcDAFrame** [comment]
## This is the decision that I'm working on right now

How to couple the repo with a form

### 1. pass it to the form view
**pros**
- reduces coupling between form view/model and repo.
**cons**
- the issue with this is that the form is then responsibile for passing the data

```jsx
<Form form={form} repo={repo}>
    ...
</Form>
```

### 2. pass it to the form model
**pros** 
- cleanest way of doing this
**cons**
- would require figuring out how to pass data to a repo
    - perhaps the `call()` function could be `call(data: any)`
```ts
this.form = new FormModel({
    submit: this.repos.submit 
})
```

**Comment by waleedza99** [comment]
**October 8th Update:**
- I (@waleedza99) am still blocked from integrating @MarcDAFrame package. There were no checks done in ensuring that the branch that @MarcDAFrame was working on was actually free of errors and ready to be merged into my branch (waleed) before declaring that I (@waleedza99) was unblocked on 2021-10-04. I have merged @MarcDAFrame branch a few times before and today (2021-10-08) as well, but every time I merged there were errors that prevented the app from simply running. This has come in the way of my progress on other tasks as each time I merged @MarcDAFrame's branch I had to spend time trying to fix the errors that came from @MarcDAFrame's branch.

**Comment by saidghamra** [comment]
> **October 8th Update:**
> - I (@waleedza99) am still blocked from integrating @MarcDAFrame package. There were no checks done in ensuring that the branch that @MarcDAFrame was working on was actually free of errors and ready to be merged into my branch (waleed) before declaring that I (@waleedza99) was unblocked on 2021-10-04. I have merged @MarcDAFrame branch a few times before and today (2021-10-08) as well, but every time I merged there were errors that prevented the app from simply running. This has come in the way of my progress on other tasks as each time I merged @MarcDAFrame's branch I had to spend time trying to fix the errors that came from @MarcDAFrame's branch. 

@MarcDAFrame Can you please look into the build errors and make sure that the app runs? @waleedza99 has sent some pictures of the errors he's running into on Discord.

**Comment by MarcDAFrame** [comment]
# Todo
- [ ] imrpove the route guards
    - perhaps use @mars-man/components loaded guard
- [ ] wiki pages
- [ ] snackbar
- [x] redirect 
    - to login if not logged in
    - to workspace if logged in

# Bugs
- Signup flow is broken, doesn't work without a workspace
    - everything else in the signup flow works pretty well
- Permissions isn't submitting right
    - empty list of permissions sent for some reason, you have to reselect
- Proxy Page unauthorized (in development due to cookies not being rewritten for that domain?)


# QOL Changes
- [ ] creates a `routes.tsx` file for all the resources (workspaces, deployments ... in the same way that Wiki and Admin have this)
- [ ] remove declarations from `index.ts(x)` files, use these just for exports
- [ ] shorten the imports (i.e. `app/res/Admin/views/Users.ts` should be `app/res/Admin`)
- [ ] remove the `src/app/OLD` folder 
- [ ] improve file names to increase file searchability (i.e. instead of `app/res/Admin/Users.tsx` use `app/res/Admin/AdminUsersTable.tsx`

**Comment by saidghamra** [comment]
The last part of this issue is integrating the mars-man/models package into projects, project lanes and project tasks.
- [ ] Projects
- [ ] Project lanes
- [ ] Project tasks

**Comment by saidghamra** [comment]
We will tackle the projects integration in another story 🎉
